src/core/namespace.c

```diff
@@ -287,12 +288,31 @@ static int apply_mount(
 
                 /* First, get rid of everything that is below if there
                  * is anything... Then, overmount it with an
-                 * inaccessible directory. */
+                 * inaccessible path. */
                 umount_recursive(m->path, 0);
 
-                what = "/run/systemd/inaccessible";
-                break;
+                /* mount() always follow symlinks,
+                 * so we follow symlinks during the stat() as well */
+                r = stat(m->path, &target);
+                if (r != 0)
+                        return -errno;
 
+                if (S_ISREG(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_REG;
+                else if (S_ISDIR(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_DIR;
+                else if (S_ISCHR(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_CHR;
+                else if (S_ISBLK(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_BLK;
+                else if (S_ISFIFO(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_FIFO;
+                else if (S_ISSOCK(target.st_mode))
+                        what = SD_INACCESSIBLE_PATH_SOCK;
+                else {
```

Comment:
I think the above if checks should be turned into a function of its own. Could even be moved into mount-util.h or so..

And you could use switch() for this, if you like, like this:

```c
static const char *mode_to_inaccessible_node(mode_t) {

        switch(target.st_mode & S_IFMT) {

        case S_IFREG:
                return "/run/systemd/inaccessible/reg";

        case S_IFDIR:
                return "/run/systemd/inaccessible/dir";
        …
        }

        return NULL;
}
```

But that's up to you. The switch is probably nicer to the compiler, but this isn't performance sensitive anyway...